pkg/sql/sessioninit/cache.go

```diff
@@ -159,15 +168,35 @@ func (a *Cache) GetAuthInfo(
 				return nil
 			}
 
-			// Lookup the data outside the lock.
-			aInfo, err = readFromSystemTables(
-				ctx,
-				txn,
-				ie,
-				username,
+			// Lookup the data outside the lock, with at most one request in-flight
+			// for each user. The user and role_options table versions are also part
+			// of the request key so that we don't read data from an old version
+			// of either table.
+			ch, _ := a.populateCacheGroup.DoChan(
+				fmt.Sprintf("%s-%d-%d", username.Normalized(), usersTableVersion, roleOptionsTableVersion),
+				func() (interface{}, error) {
+					// Use a different context to fetch, so that it isn't possible for
+					// one query to timeout and cause all the goroutines that are waiting
+					// to get a timeout error.
+					ctx, cancel := a.stopper.WithCancelOnQuiesce(
+						logtags.WithTags(context.Background(), logtags.FromContext(ctx)))
+					defer cancel()
+					return readFromSystemTables(
+						ctx,
+						txn,
+						ie,
+						username,
+					)
+				},
 			)
-			if err != nil {
-				return err
+			select {
+			case res := <-ch:
+				if res.Err != nil {
+					return res.Err
+				}
+				aInfo = res.Val.(AuthInfo)
+			case <-ctx.Done():
+				return ctx.Err()
```

Comment:
should we wrap this error with `error during authentication` or something?

Reply:
That already happens here: https://github.com/cockroachdb/cockroach/blob/fc7b49fd83892044e8c7c21b8b71888c383141d0/pkg/sql/user.go#L238

and there's a test for it: https://github.com/cockroachdb/cockroach/blob/fc7b49fd83892044e8c7c21b8b71888c383141d0/pkg/sql/user_test.go#L178